sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,139 @@
+
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+import org.apache.spark.sql.execution.QueryExecution
+import org.apache.spark.sql.util.QueryExecutionListener
+
+/**
+ * Not thread-safe.
+ * @param name
+ * @param sparkSession
+ */
+class Observation(name: String) {
+
+  private val listener: ObservationListener = ObservationListener(this)
+
+  private var sparkSession: Option[SparkSession] = None
+
+  @volatile private var row: Option[Row] = None
+
+  /**
+   * Attach this observation to the given Dataset.
+   *
+   * @param ds dataset
+   * @tparam T dataset type
+   * @return observed dataset
+   */
+  def on[T](ds: Dataset[T])(expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Wait for the first action on the observed dataset to complete and returns true.
+   * This method times out after the given amount of time and returns false.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action complete within timeout, false on timeout
+   */
+  def waitCompleted(time: Long, unit: TimeUnit): Boolean = waitCompleted(Some(time), unit)
+
+  /**
+   * Get the observation results. This waits until the observed dataset finishes its first action.
+   * If you want to wait for the result and provide a timeout, use waitCompleted.
+   * Only the result of the first action is available. Subsequent actions do not modify the result.
+   */
+  def get: Row = {
+    assert(waitCompleted(None, TimeUnit.SECONDS), "waitCompleted without timeout returned false")
+    row.get
+  }
+
+  private def waitCompleted(time: Option[Long], unit: TimeUnit): Boolean = {
+    synchronized {
+      if (row.isEmpty) {
+        if (time.isDefined) {
+          this.wait(unit.toMillis(time.get))
+        } else {
+          this.wait()
+        }
+      }
+      row.isDefined
+    }
+  }
+
+  private def register(sparkSession: SparkSession): Unit = {
+    // makes this class thread-safe:
+    // only the first thread entering this block can set sparkSession
+    // all other threads will see the exception, because it is only allowed to do this once
+    synchronized {
+      if (this.sparkSession.isDefined) {
+        throw new IllegalStateException("An Observation can be used with a Dataset only once")
+      }
+      this.sparkSession = Some(sparkSession)
+    }
+
+    sparkSession.listenerManager.register(listener)
+  }
+
+  private def unregister(): Unit = {
+    this.sparkSession.foreach(_.listenerManager.unregister(listener))
+  }
+
+  private[spark] def onFinish(qe: QueryExecution): Unit = {
+    synchronized {
+      this.row = qe.observedMetrics.get(name)
+      assert(this.row.isDefined, "No metric provided by QueryExecutionListener")
+      this.notifyAll()
+    }
+    unregister()
```

Comment:
shall we put it in the synchronized block? `sparkSession` is updated by another thread in `register`

Reply:
BTW `unregister` is just one line code, seems we don't need this method and can just write `this.sparkSession.foreach(_.listenerManager.unregister(listener))` here.

Reply:
> shall we put it in the synchronized block? ``

This means we would guard against another thread that is concurrently running `register`. The listener thread can only call into `unregister` after `register` has finished. So yes, two threads are involved in updating and accessing `sparkSession`, but not concurrently.

Reply:
> BTW `unregister` is just one line code

Yes, that one-liner could be place instead of `unregister()`, but that would not improve readability.

Right now, in `onFinish` we exactly know what is happening at the last line: unregister. With that one-liner instead, that is not immediately clear.

With `unregister()` we say what to do and with `this.sparkSession.foreach(_.listenerManager.unregister(listener))` we would say how to do it and everyone has to encode what it means.

Reply:
> two threads are involved in updating and accessing sparkSession, but not concurrently.

My understanding is that, even if two things can't happen concurrently, we still need to add `volatile` or do lock as long as the update and read of the value happens in different threads. Otherwise, it's non-deterministic that how soon other threads can see the value update.

Reply:
You are saying making `sparkSession` needs to be `volatile` because it is written and read in different threads, even if not concurrently. That is right, I'l fix that.

But reading and writing it does not need to be in a `synchronized` because that is not happening concurrently and thus cannot interfere.